crucible-symio/src/Lang/Crucible/SymIO.hs

```diff
@@ -0,0 +1,790 @@
+    liftIO $ addFailedAssertion sym $ AssertFailureSimError "Failed to write array due to closed file handle." "writeArray': Closed file handle"
+  Right sz' -> return sz'
+
+-- | Read a byte from a given 'FileHandle' and increment it.
+-- The partial result is undefined if the read yields no results.
+readByte ::
+  (IsSymInterface sym, 1 <= wptr) =>
+  GlobalVar (FileSystemType wptr) ->
+  FileHandle sym wptr ->
+  (forall args'. Either FileHandleError (PartExpr (W4.Pred sym) (W4.SymBV sym 8)) -> C.OverrideSim p sym arch r args' ret a) ->
+  C.OverrideSim p sym arch r args ret a
+readByte fsVar fhdl cont = runFileMHandleCont fsVar fhdl emptyRegMap cont $ \_ fhdl' -> do
+  ptr <- getHandle fhdl'
+  v <- readBytePointer ptr
+  sym <- getSym
+  repr <- getPtrSz
+  one <- liftIO $ W4.bvLit sym repr (BV.mkBV repr 1)
+  readBytes <- incHandleRead fhdl' one
+  valid <- liftIO $ W4.isEq sym one readBytes
+  return $ mkPE valid v
+
+-- | Total version of 'readByte' that asserts success.
+readByte' ::
+  (IsSymInterface sym, 1 <= wptr) =>
+  GlobalVar (FileSystemType wptr) ->
+  FileHandle sym wptr ->
+  C.OverrideSim p sym arch r args ret (PartExpr (W4.Pred sym) (W4.SymBV sym 8))
+readByte' fsVar fhdl = readByte fsVar fhdl $ \case
+  Left FileHandleClosed -> do
+    sym <- C.getSymInterface
+    liftIO $ addFailedAssertion sym $ AssertFailureSimError "Failed to read byte due to closed file handle." "readByte': Closed file handle"
+  Right r -> return r
+
+-- | Read a chunk from a given 'FileHandle' of the given size, and increment the
+-- handle by the size. Returns a struct containing the array contents, and the number
+-- of bytes read.
+readChunk ::
+  (IsSymInterface sym, 1 <= wptr) =>
+  GlobalVar (FileSystemType wptr) ->
+  FileHandle sym wptr ->
+  W4.SymBV sym wptr ->
+  (forall args'. Either FileHandleError (DataChunk sym wptr, W4.SymBV sym wptr) -> C.OverrideSim p sym arch r args' ret a) ->
+  C.OverrideSim p sym arch r args ret a
+readChunk fsVar fhdl sz cont = do
+  W4.BaseBVRepr ptrSz <- return $ W4.exprType sz
+  let args = RegMap (Empty :> RegEntry (BVRepr ptrSz) sz)
+  runFileMHandleCont fsVar fhdl args cont $ \(RegMap (Empty :> RegEntry _ sz')) fhdl' -> do
+    ptr <- getHandle fhdl'
+    chunk <- readChunkPointer ptr sz'
+    readSz <- incHandleRead fhdl' sz'
+    return (chunk, readSz)
+
+-- | Partial version of 'readArray' that asserts success.
+readChunk' ::
+  (IsSymInterface sym, 1 <= wptr) =>
+  GlobalVar (FileSystemType wptr) ->
+  FileHandle sym wptr ->
+  W4.SymBV sym wptr ->
+  C.OverrideSim p sym arch r args ret (DataChunk sym wptr, W4.SymBV sym wptr)
+readChunk' fsVar fhdl sz = readChunk fsVar fhdl sz $ \case
+  Left FileHandleClosed -> do
+    sym <- C.getSymInterface
+    liftIO $ addFailedAssertion sym $ AssertFailureSimError "Failed to read array due to closed file handle." "readArray': Closed file handle"
+  Right arr -> return arr
+
+-- | Returns a predicate indicating whether or not the file handle
```

Comment:
Is this comment correct? It mentions EOF, but the function seems to be checking if the handle is open.

Reply:
Also going to refer this one to @danmatichuk. I'm also curious, reading around there, if we might be conflating EOF and closed. EOF can change if more data comes in, while being closed cannot.

Reply:
Or if you seek, etc.  We should definitely keep track of EOF separately from closed.

Reply:
Right, this comment is wrong (at some point they were conflated, but that was fixed). This is checking that the handle is still open. If a handle reaches EOF it can still be read from, it will just return that zero bytes have been read.

Reply:
I'll update the comment